[Pipelining] Relax scale_grads assert

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #144834
* #145099
* #145011
* __->__ #145010

The assert felt morally valid- if no gradients are scaled, then something
is definitely wrong with the setup.  In one instance, PP +
optimizer-in-backward (in torchtitan) resulted in grad=None after
running .backward() and before scaling grads.

On the other hand, the existing assert is too restrictive.  It's
possible that a model used with pipelining would have some parameters
that do not receieve gradients, and we shouldn't hard-error in these
cases.  (E.g. if the parameter is literally not used, or is frozen).
In the extreme case, the whole stage could be frozen.  So we do not
complain if no grads are scaled.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

wconstab committed:
[Pipelining] Relax scale_grads assert

The assert is morally valid- if no gradients are scaled, then something
is definitely wrong with the setup.  In one instance, PP +
optimizer-in-backward (in torchtitan) resulted in grad=None after
running .backward() and before scaling grads.  This is obviously not a
correct scenario.

On the other hand, the existing assert is too restrictive.  It's
possible that a model used with pipelining would have some parameters
that do not receieve gradients, and we shouldn't hard-error in these
cases.  (E.g. if the parameter is literally not used, or is frozen).

[ghstack-poisoned]

mori360 reviewed (approved):
Thanks for the PR! LGTM

tianyu-l commented on torch/distributed/pipelining/stage.py:
```diff
@@ -586,9 +586,15 @@ def scale_grads(self, grad_scale_factor: int) -> None:
+                if p.grad is not None:
+                    p.grad.div_(grad_scale_factor)
+                    scaled_any = True
+            assert scaled_any, (
```
might be a corner case:
What if I have a giant model, some parts frozen and some parts being trained, and on one of the stage I have to put a entirely frozen component.

wconstab replied:
that's a good point. i think i can just delete this assertion entirely, although it did catch a composability bug.

In any case, this codepath should go away.  @H-Huang is switching pipelining to use .backward, after which point we can use hooks, and then I think we should do grad scaling via hooks instead of loops over parameters.

So maybe I just land it for now and then we delete it when we migrate?

tianyu-l replied:
sgtm, except that I think we should understand the compile+backward hook status before doing so, including why optimizer-in-backward using backward hooks would not fail `torch.compile(fullgraph=True)` but fails when CA is used.

wconstab committed:
Update on "[Pipelining] Relax scale_grads assert"


The assert is morally valid- if no gradients are scaled, then something
is definitely wrong with the setup.  In one instance, PP +
optimizer-in-backward (in torchtitan) resulted in grad=None after
running .backward() and before scaling grads.  This is obviously not a
correct scenario.

On the other hand, the existing assert is too restrictive.  It's
possible that a model used with pipelining would have some parameters
that do not receieve gradients, and we shouldn't hard-error in these
cases.  (E.g. if the parameter is literally not used, or is frozen).

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[Pipelining] Relax scale_grads assert"


The assert felt morally valid- if no gradients are scaled, then something
is definitely wrong with the setup.  In one instance, PP +
optimizer-in-backward (in torchtitan) resulted in grad=None after
running .backward() and before scaling grads.

On the other hand, the existing assert is too restrictive.  It's
possible that a model used with pipelining would have some parameters
that do not receieve gradients, and we shouldn't hard-error in these
cases.  (E.g. if the parameter is literally not used, or is frozen).
In the extreme case, the whole stage could be frozen.  So we do not
complain if no grads are scaled.

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
